### PR TITLE
fix Issue 17349 - Covariant overrides should be allowed

### PR DESCRIPTION
**Pull request by WalterBright**
This is important to making Object's methods @safe.

Fix Issue https://issues.dlang.org/show_bug.cgi?id=16303

Fix Issue https://issues.dlang.org/show_bug.cgi?id=17354

Fix Issue https://issues.dlang.org/show_bug.cgi?id=16600

Fix Issue https://issues.dlang.org/show_bug.cgi?id=16538

### CODE REVIEW

**Comment by WalterBright** [comment]
This is giving a jenkins error of:
```
../../../../../.dub/packages/libdparse-0.7.0-beta.2/libdparse/src/dparse/ast.d(1344,10): Error: cannot implicitly override base class method object.Object.opEquals with dparse.ast.Declaration.opEquals; add 'override' attribute
```

File: test/compilable/fix17349.d [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+class D : C {
+    override void bar() const;
+    override void foo(const void*);
+    override void abc(const Object);
+    override void def(const S);
+}
+
+alias fp_t = void function(int*);
+void abc(const int*);
+fp_t fp = &abc;
```
Comment (mathias-lang-sociomantic) [inline test/compilable/fix17349.d]: Doesn't that fix https://issues.dlang.org/show_bug.cgi?id=16303 also ?
Could you add a test case with delegates ?

Comment (WalterBright) [inline test/compilable/fix17349.d]: yes, thanks for the tip.

**Comment by WalterBright** [comment]
Ok, got it down to this failure on vibe.d:
```
../../.dub/packages/libasync-0.8.3/libasync/source/libasync/posix2.d(789,5): Error: memutils.hashmap.HashMap!(Tuple!(int, uint), DWFolderInfo, ThreadMem).HashMap.opApply called with argument types (int delegate(ref const(Tuple!(int, uint)) key, ref const(DWFolderInfo) info) nothrow @system) matches both:
../../.dub/packages/memutils-0.4.9/memutils/source/memutils/hashmap.d(151,6):     memutils.hashmap.HashMap!(Tuple!(int, uint), DWFolderInfo, ThreadMem).HashMap.opApply(int delegate(ref const(Tuple!(int, uint)), ref DWFolderInfo) del)
and:
../../.dub/packages/memutils-0.4.9/memutils/source/memutils/hashmap.d(160,6):     memutils.hashmap.HashMap!(Tuple!(int, uint), DWFolderInfo, ThreadMem).HashMap.opApply(int delegate(ref const(Tuple!(int, uint)), ref const(DWFolderInfo)) del) const
```

**Comment by BBasile** [comment]
This PR and (through the the Jenkins failure) actually reveals this issue: https://issues.dlang.org/show_bug.cgi?id=17354. As far as i understand it would mean that this PR also fixes the bug that was that a case of hijack was not detected in final classes.

**Comment by andralex** [comment]
I think we should keep the issue open until we implement thunks, too.

**Comment by WalterBright** [comment]
This is ready to go now. (The continuous- error is not relevant.)

File: src/ddmd/func.d [hunk 2]
```diff
@@ -778,9 +778,16 @@ extern (C++) class FuncDeclaration : Declaration
                     if (fdv.isFinalFunc())
                         error("cannot override final function %s", fdv.toPrettyChars());
 
-                    doesoverride = true;
                     if (!isOverride())
-                        .error(loc, "cannot implicitly override base class method %s with %s; add 'override' attribute", fdv.toPrettyChars(), toPrettyChars());
+                    {
+                            int vi2 = findVtblIndex(&cd.baseClass.vtbl, cast(int)cd.baseClass.vtbl.dim, false);
+                            if (vi2 < 0)
+                                .deprecation(loc, "fix17349 cannot implicitly override base class method %s with %s; add 'override' attribute", fdv.toPrettyChars(), toPrettyChars());
```
Comment (mathias-lang-sociomantic) [inline src/ddmd/func.d]: Extra `fix17349`

File: src/ddmd/mtype.d [hunk 3]
```diff
@@ -777,7 +809,7 @@ extern (C++) abstract class Type : RootObject
                 if (!t2.isscope)
                     stc2 |= STCref;
             }
-            if (!Parameter.isCovariantScope(t1.isref, stc1, stc2))
+            if (global.params.vsafe && !Parameter.isCovariantScope(t1.isref, stc1, stc2))
```
Comment (mathias-lang-sociomantic) [inline src/ddmd/mtype.d]: That looks unrelated. Isn't it another bugfix ?

File: src/ddmd/func.d (hunk 2 above)
Comment (WalterBright) [inline src/ddmd/func.d]: ??

File: src/ddmd/func.d [hunk 4]
```diff
@@ -2286,12 +2293,15 @@ extern (C++) class FuncDeclaration : Declaration
...
      * this function overrides.
      * Prefer an exact match to a covariant one.
+     * Params:
+     *      fix17349 = enable fix https://issues.dlang.org/show_bug.cgi?id=17349
      * Returns:
      *      -1      didn't find one
      *      -2      can't determine because of forward references
      */
-    final int findVtblIndex(Dsymbols* vtbl, int dim)
+    final int findVtblIndex(Dsymbols* vtbl, int dim, bool fix17349 = true)
```
Comment (yebblies) [inline src/ddmd/func.d]: Is this going to be permanent? If so if needs a much better name that actually describes what it does.

File: test/runnable/sctor.d [hunk 5]
```diff
@@ -360,7 +360,7 @@ void test14450()
 
     { auto i = immutable S14450a(1);    assert(i.x == 2); }
     { auto i = immutable S14450b(1);    assert(i.x == 2); }
-    { auto i = immutable S14450c!()(1); assert(i.x == 2); }
+    { auto i = immutable S14450c!()(1); assert(i.x == 1); }
```
Comment (yebblies) [inline test/runnable/sctor.d]: Isn't this wrong? It looks like the wrong overload will get called

File: src/ddmd/func.d (hunk 4 above)
Comment (WalterBright) [inline src/ddmd/func.d]: I thought about it, and couldn't think of anything better for the time being. (The name does provide a reference to what it does.)

File: test/runnable/sctor.d (hunk 5 above)
Comment (WalterBright) [inline test/runnable/sctor.d]: Actually, it brings it in line with the behavior a few lines above with the `== 1` code. Those cases are all non-template versions of the constructors.

File: src/ddmd/func.d (hunk 2 above)
Comment (mathias-lang-sociomantic) [inline src/ddmd/func.d]: Why does "fix17349" appears in the error message provided to the user ?

File: test/runnable/sctor.d (hunk 5 above)
Comment (yebblies) [inline test/runnable/sctor.d]: But it's calling the mutable ctor on an immutable struct? Am i missing something here?

File: src/ddmd/func.d (hunk 4 above)
Comment (yebblies) [inline src/ddmd/func.d]: It would be nice if we could avoid the indirection - `enableContravariance17349` maybe?

File: test/runnable/sctor.d (hunk 5 above)
Comment (WalterBright) [inline test/runnable/sctor.d]: Look at the `==1` cases above. It's the same.

File: src/ddmd/func.d (hunk 4 above)
Comment (WalterBright) [inline src/ddmd/func.d]: I don't think the longer name adds much. There's a clickable link in the description of the parameter that should suffice. I think the dmd source code in general could benefit more from clickable links in the comments :-)

**Comment by WalterBright** [comment]
This is blocking https://github.com/dlang/phobos/pull/5387

**Comment by wilzbach** [comment]
@WalterBright FYI the [dlang-bot supports multiple issues](https://github.com/dlang-bots/dlang-bot), e.g.

> fix issue 16303, 16538, 16600, 17349, 17354 - Covariant overrides should be allowed

 this [can also be easily checked online](https://regex101.com/r/aI0Rp6/4).

**Comment by WalterBright** [comment]
Why doesn't it work with the first comment?

**Comment by CyberShadow** [comment]
I believe it looks at the commit message.

You can add "Fixes issues 16303, 16538, 16600, 17349, 17354." in the commit message body.

**Comment by wilzbach** [comment]
> I believe it looks at the commit message.

Yes, (as explained on its docs), the main purpose of the bot is to ensure that issues are properly tagged in the git log, from which the changelog is generated. Other stuff like the bot auto-closing issues is just convenience build on top of it ;-)

File: src/ddmd/func.d (hunk 4 above)
Comment (yebblies) [inline src/ddmd/func.d]: I guess it's not too bad if it's non-permanent - and it seems to be used only for detecting and displaying the deprecation message.  In general though, a clickable link is NOT a substitute for explaining the purpose of the parameter at least in the ddoc.

File: test/compilable/fix16600.d [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    int i;
+
+    this(string) { i = 1; }
+    this(string) immutable { i = 2; }
+}
+
+void main()
+{
+    auto a = const(S)("abc");
+    assert(a.i == 2);
```
Comment (yebblies) [inline test/compilable/fix16600.d]: Why, why is immutable preferred over mutable here?  They both convert to const, why isn't this ambiguous?

File: test/runnable/sctor.d (hunk 5 above)
Comment (yebblies) [inline test/runnable/sctor.d]: The `==1` cases above are making a `S14450c!()`, this case is making an `immutable S14450c!()` - so they're not the same.  Why wouldn't this use the immutable ctor?

**Comment by WalterBright** [comment]
ping @yebblies @mathias-lang-sociomantic

File: src/ddmd/func.d [hunk 7]
```diff
@@ -779,9 +779,17 @@ extern (C++) class FuncDeclaration : Declaration
                     if (fdv.isFinalFunc())
                         error("cannot override final function %s", fdv.toPrettyChars());
 
-                    doesoverride = true;
                     if (!isOverride())
-                        .error(loc, "cannot implicitly override base class method %s with %s; add 'override' attribute", fdv.toPrettyChars(), toPrettyChars());
+                    {
+                            int vi2 = findVtblIndex(&cd.baseClass.vtbl, cast(int)cd.baseClass.vtbl.dim, false);
```
Comment (andralex) [inline src/ddmd/func.d]: Pardon my nitpicking... this looks like a double indent.

File: src/ddmd/func.d (hunk 4 above)
Comment (andralex) [inline src/ddmd/func.d]: Is this a blocker?

File: test/fail_compilation/fail16600.d [hunk 8]
```diff
@@ -1,21 +1,26 @@
...
+    int i;
+
+    this(string) { i = 1; }
+    this(string) immutable { i = 2; }
 }
 
 void main()
 {
     auto a = const(S)("abc");
+    assert(a.i == 2);
```
Comment (andralex) [inline test/fail_compilation/fail16600.d]: ah, finally

File: test/fail_compilation/fail_scope.d [hunk 9]
```diff
@@ -25,6 +21,10 @@ fail_compilation/fail_scope.d(137): Error: returning `foo16226(i)` escapes a ref
 //fail_compilation/fail_scope.d(40): Error: scope variable p may not be returned
 */
 
+
+
+
+
```
Comment (andralex) [inline test/fail_compilation/fail_scope.d]: why them empty lines (not a blocker)

**Review by andralex — APPROVED** [review]
Spec pull please? I can help with that.

**Comment by andralex** [comment]
Another thing: I don't see a test specifically for adding `@safe`.

File: src/ddmd/func.d (hunk 4 above)
Comment (WalterBright) [inline src/ddmd/func.d]: no

File: src/ddmd/func.d (hunk 7 above)
Comment (WalterBright) [inline src/ddmd/func.d]: That comes from interaction with the @__future code.

File: test/fail_compilation/fail_scope.d (hunk 9 above)
Comment (WalterBright) [inline test/fail_compilation/fail_scope.d]: I add them sometimes so I can add test cases in TEST_OUTPUT without having to renumber them. I learned that when programming in BASIC :-)

**Comment by WalterBright** [comment]
> a test specifically for adding @safe.

Added.

**Comment by RazvanN7** [comment]
This PR has introduced a regression: https://issues.dlang.org/show_bug.cgi?id=21538
